### DISCUSSION TITLE
被xmake的缓存困扰

### DISCUSSION THREAD

**Discussion by heheda123123** [xmake-io/xmake#4471] [General]
xmake在不同路径下执行，最终的效果不一样

https://gitee.com/scu319hy/edx/issues/I8LTIE#note_23260909_link

**Comment by heheda123123** [comment]
这是xmake的缓存机制造成的吗？只要命令不变，即便删除了项目目录下的.xmake，也会得到一样的包版本？
缓存放在用户目录下的.xmake里面？该怎么清理呢？

**Comment by heheda123123** [comment]
已解决，加-c参数即可
不过这个缓存机制着实有点坑，一般都是把缓存记录到项目目录下的某个文件吧。
这里把项目下的配置文件都删了也解决不了。

**Reply by waruqi** [reply]
所有缓存都在 当前项目的 .xmake 你删它 跟执行 `xmake f -c` 效果一样的。。。没其他地方了。

**Reply by heheda123123** [reply]
那好奇怪，你看我上面的链接，如果是执行的原来的命令就始终用的0.1版本的cpptrace库。只有把用户目录下的.xmake也删掉才会用新版本。
不是故意设计的话，可能是bug

**Reply by waruqi** [reply]
那是包，包默认全局安装的，跟缓存没有关系。。~/.xmake/packages 是包的全局安装目录，不是缓存。。

你更新了包配置，就执行 `xrepo update-repo` 更新仓库，然后执行 `xmake f -c` 重新检测包，如果包版本有更新会重新安装，如果包版本没变，不会重装

**Reply by heheda123123** [reply]
我不是说包的版本，是说执行xmake project -k ninja生成的build.ninja里面引用的包版本仍然是旧的

**Reply by waruqi** [reply]
那就是本地缓存问题，执行 `xmake f -c` 重新检测下包版本就好了，这跟删 `.xmake` 效果一样的。。

**Reply by heheda123123** [reply]
链接里面就是提到了，手动删掉项目路径下的.xmake没效果

**Reply by heheda123123** [reply]
我把完整复现过程列在下面了，那有没有可能是xmake会在执行xmake的这个目录进行缓存呢？因为下面的复现过程是用了-P参数在非项目目录下操作的

**Comment by heheda123123** [comment]
不用edx，完全手动执行xmake命令也能复现
1 首先准备环境，此时本地xmake-repo仓库里还是cpptrace 0.1版本
2 在这个路径下创建一个标准的xmake项目 `/home/cat/.edx/c/Users/Administrator/temp/edx/testxxx`
3 xmake.lua里面加入cpptrace的依赖，src/main.cpp里面加入使用cpptrace的代码
4 然后在用户目录下执行下面的配置命令
```
xmake f -y -P ".edx/c/Users/Administrator/temp/edx/testxxx" --buildir=".edx/c/Users/Administrator/temp/edx/testxxx/.edx/xmake/cat@192.168.181.128(gcc-x86_64-linux-gnu-11.3.0)/Debug" -p linux --cc="/usr/bin/x86_64-linux-gnu-gcc-11" --cxx="/usr/bin/x86_64-linux-gnu-g++-11" --arch=x86_64 --mode=debug
```
5 生成compile_commands.json
```
xmake project -y -P ".edx/c/Users/Administrator/temp/edx/testxxx" -k compile_commands --lsp=clangd --archs="x86_64" --modes="debug" ".edx/xmake/cat@192.168.181.128(gcc-x86_64-linux-gnu-11.3.0)/Debug"
```
6 生成build.ninja
```
xmake project -y -P ".edx/c/Users/Administrator/temp/edx/testxxx" -k ninja --archs="x86_64" --modes="debug" ".edx/xmake/cat@192.168.181.128(gcc-x86_64-linux-gnu-11.3.0)/Debug"
```
7 构建项目,因为前面用了cpptrace0.3.1的代码，所以肯定会失败
```
xmake build -y -P ".edx/c/Users/Administrator/temp/edx/testxxx" testxxx
```
8 更新本地的xrepo仓库
```
xrepo update-repo
```
9 删掉项目目录下的.edx和.xmake目录
```
cat@zoo:~/.edx/c/Users/Administrator/temp/edx/testxxx$ rm -rf .edx/
cat@zoo:~/.edx/c/Users/Administrator/temp/edx/testxxx$ rm -rf .xmake/
```
10 重新配置
```
xmake f -y -P ".edx/c/Users/Administrator/temp/edx/testxxx" --buildir=".edx/c/Users/Administrator/temp/edx/testxxx/.edx/xmake/cat@192.168.181.128(gcc-x86_64-linux-gnu-11.3.0)/Debug" -p linux --cc="/usr/bin/x86_64-linux-gnu-gcc-11" --cxx="/usr/bin/x86_64-linux-gnu-g++-11" --arch=x86_64 --mode=debug
```
11 重新生成build.ninja
```
xmake project -y -P ".edx/c/Users/Administrator/temp/edx/testxxx" -k ninja --archs="x86_64" --modes="debug" ".edx/xmake/cat@192.168.181.128(gcc-x86_64-linux-gnu-11.3.0)/Debug"
```
12 发现生成的build.ninja里面仍然是用的 0.1版本的cpptrace

相关操作截图如下，带了时间的
![image](https://github.com/xmake-io/xmake/assets/97490782/a7519b8b-1155-437f-b46d-5b986d0bafec)
![image](https://github.com/xmake-io/xmake/assets/97490782/c5e424e4-f5a0-4e00-a647-1549f47d2f0a)

**Comment by heheda123123** [comment]
在第一幅图第61行，第二次执行配置的时候就有点不对劲，我这里加了vD参数，尽管我删掉了项目testxxx路径下的.edx目录。
但是可以看到瞬间就返回了
![image](https://github.com/xmake-io/xmake/assets/97490782/0083c679-291e-4659-960e-6964ea006d87)

**Comment by heheda123123** [comment]
我好像知道为啥了，因为xmake f 的时候读取和生成 .xmake 不是在项目路径下生成，而是在当前执行xmake命令所在目录生成。
这就导致尽管我上面删除了构建目录，但是因为我是在用户目录下通过-P参数指定的，.xmake生成在了用户目录下
这算是bug吧，项目的配置应该还是写在项目里比较合适

**Comment by waruqi** [comment]
> xmake f -y -P ".edx/c/Users/Administrator/temp/edx/testxxx"
> cat@zoo:~/.edx/c/Users/Administrator/temp/edx/testxxx$ rm -rf .edx/
> cat@zoo:~/.edx/c/Users/Administrator/temp/edx/testxxx$ rm -rf .xmake/

你看下 https://github.com/xmake-io/xmake/issues/3342 这个对 `xmake f -P ` 的描述

2.7.7 之后增加了对 workdir 的锁定，projectdir 外置编译支持。`xmake f -P xxx` 配置后，.xmake 是在当前目录的，不是 xxx 目录下，xxx 会保证完整干净。。

<img width="907" alt="image" src="https://github.com/xmake-io/xmake/assets/151335/9d2dd3c8-bbdc-44a2-ac52-58b6d23ed5a2">


另外，只需要设置一次 `-P` 后面的所有命令，就不需要额外加 `-P` 的

**Reply by heheda123123** [reply]
啊这，下意识以为项目的配置在项目下。有意这样设计的那就算了

**Reply by waruqi** [reply]
所以不建议直接删 .xmake 。那个本来就是内部文件，不对用户开放，走 `xmake f -c` 就行了，`-c` 会保证重新检测

**Reply by heheda123123** [reply]
好的，谢谢

**Comment by heheda123123** [comment]
对了大佬，这里提到另外一个问题。当执行xmake f 配置的时候，是否应该检查一下当前配置里的包版本和本地 xmake-repo 列表里面的版本
我之前看了xmake的文档，也只是提到专门加了包依赖锁定才不会升级
https://gitee.com/scu319hy/edx/issues/I8LTIE#note_23272579_link

**Reply by waruqi** [reply]
按需自己开 require lock
